python/pywatchman/__init__.py

```diff
@@ -114,16 +126,128 @@ def __init__(self):
 
     LocalFree = ctypes.windll.kernel32.LocalFree
 
-    GetOverlappedResultEx = ctypes.windll.kernel32.GetOverlappedResultEx
-    GetOverlappedResultEx.argtypes = [wintypes.HANDLE,
-                                      ctypes.POINTER(OVERLAPPED), LPDWORD,
-                                      wintypes.DWORD, wintypes.BOOL]
-    GetOverlappedResultEx.restype = wintypes.BOOL
+    # Windows 7 and earlier does not support GetOverlappedResultEx
+    # The alternative is to use GetOverlappedResult and wait
+    # for read or write operation to complete. This is
+    # done be using CreateEvent and WaitForSingleObject.
+    # Windows 8 is version 6.2 - see the Operating System Version
+    # page on MSDN
+    WINDOWS_8_AND_LATER = sys.getwindowsversion() >= (6, 2)
+    if not WINDOWS_8_AND_LATER:
+        CreateEvent = ctypes.windll.kernel32.CreateEventA
+        CreateEvent.argtypes = [LPDWORD, wintypes.BOOL, wintypes.BOOL,
+                                wintypes.LPSTR]
+        CreateEvent.restype = wintypes.HANDLE
+
+        GetOverlappedResult = ctypes.windll.kernel32.GetOverlappedResult
+        GetOverlappedResult.argtypes = [wintypes.HANDLE,
+                                        ctypes.POINTER(OVERLAPPED), LPDWORD,
+                                        wintypes.BOOL]
+        GetOverlappedResult.restype = wintypes.BOOL
+
+        WaitForSingleObject = ctypes.windll.kernel32.WaitForSingleObject
+        WaitForSingleObject.argtypes = [wintypes.HANDLE, wintypes.DWORD]
+        WaitForSingleObject.restype = wintypes.DWORD
+    else:
+        GetOverlappedResultEx = ctypes.windll.kernel32.GetOverlappedResultEx
+        GetOverlappedResultEx.argtypes = [wintypes.HANDLE,
+                                          ctypes.POINTER(OVERLAPPED), LPDWORD,
+                                          wintypes.DWORD, wintypes.BOOL]
+        GetOverlappedResultEx.restype = wintypes.BOOL
```

Comment:
the approach that we took for this in the C implementation was to provide our own implementation of this function for Windows 7:

https://github.com/facebook/watchman/blob/master/stream_win.c#L49-L78

it would be easier to maintain if we could use broadly the same approach in here; that way it would be easier to see how to apply bug fixes from one to another.

It should also help to reduce the amount of conditional code throughout this module.  Would you mind taking a look at phrasing this change in that way?